app/src/extra/java/org/wikipedia/donate/GooglePayViewModel.kt

```diff
@@ -0,0 +1,150 @@
+import org.wikipedia.util.GeoUtil
+import org.wikipedia.util.Resource
+import org.wikipedia.util.log.L
+import java.text.NumberFormat
+import java.time.Instant
+import java.util.Locale
+import java.util.concurrent.TimeUnit
+import kotlin.math.abs
+
+class GooglePayViewModel : ViewModel() {
+    val uiState = MutableStateFlow(Resource<DonationConfig>())
+    var donationConfig: DonationConfig? = null
+    private val currentCountryCode: String get() = GeoUtil.geoIPCountry.orEmpty()
+
+    val currencyFormat: NumberFormat = NumberFormat.getCurrencyInstance(Locale.Builder()
+        .setRegion(currentCountryCode).setLanguage(WikipediaApp.instance.appOrSystemLanguageCode).build())
+    val currencyCode get() = currencyFormat.currency?.currencyCode ?: "USD"
+    val currencySymbol get() = currencyFormat.currency?.symbol ?: "$"
+    val decimalFormat = GooglePayComponent.getDecimalFormat(currencyCode)
+
+    val transactionFee: Float get() = donationConfig?.currencyTransactionFees?.get(currencyCode)
+        ?: donationConfig?.currencyTransactionFees?.get("default") ?: 0f
+
+    val emailOptInRequired get() = donationConfig?.countryCodeEmailOptInRequired.orEmpty().contains(currentCountryCode)
+
+    var finalAmount = 0f
+
+    init {
+        currencyFormat.minimumFractionDigits = 0
+        load()
+    }
+
+    fun load() {
+        viewModelScope.launch(CoroutineExceptionHandler { _, throwable ->
+            uiState.value = Resource.Error(throwable)
+        }) {
+            uiState.value = Resource.Loading()
+
+            val donationConfigCall = async { DonationConfigHelper.getConfig() }
+
+            donationConfig = donationConfigCall.await()
+
+            // The paymentMethods API is rate limited, so we cache it manually.
+            val now = Instant.now().epochSecond
+            if (abs(now - Prefs.paymentMethodsLastQueryTime) > TimeUnit.DAYS.toSeconds(7)) {
+                Prefs.paymentMethodsMerchantId = ""
+                Prefs.paymentMethodsGatewayId = ""
+
+                val paymentMethodsCall = async {
+                    ServiceFactory.get(WikiSite(GooglePayComponent.PAYMENTS_API_URL))
+                        .getPaymentMethods(currentCountryCode)
+                }
+                paymentMethodsCall.await().response?.let { response ->
+                    Prefs.paymentMethodsLastQueryTime = now
+                    response.paymentMethods.find { it.type == GooglePayComponent.PAYMENT_METHOD_NAME }?.let {
+                        Prefs.paymentMethodsMerchantId = it.configuration?.merchantId.orEmpty()
+                        Prefs.paymentMethodsGatewayId = it.configuration?.gatewayMerchantId.orEmpty()
+                    }
+                }
+            }
+
+            if (Prefs.paymentMethodsMerchantId.isEmpty() || Prefs.paymentMethodsGatewayId.isEmpty()) {
+                uiState.value = NoPaymentMethod()
+            } else {
+                uiState.value = Resource.Success(donationConfig!!)
```

Comment:
I'm actually not sure this would be safer...  If `donationConfig` is null, it will throw a NullPointerException, which will be caught by the `CoroutineExceptionHandler` at the top, and will automatically provide a `Resource.Error`.